cli/data.go

```diff
@@ -35,89 +37,183 @@ func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
 		return errors.Wrapf(err, "error creating destination directories")
 	}
 
-	var last string
-	numFilesDownloaded := 0
-	for {
-		var err error
-		var resp *datapb.BinaryDataByFilterResponse
-		for count := 0; count < maxRetryCount; count++ {
-			resp, err = c.sendBinaryDataByFilterRequest(filter, last)
-			if err == nil {
+	if concurrentDownloads == 0 {
+		concurrentDownloads = defaultConcurrentDownloads
+	}
+
+	ids := make(chan string, concurrentDownloads)
+	// Give channel buffer of 1+concurrentDownloads because that is the number of goroutines that may be passing an
+	// error into this channel (1 get ids routine + concurrentRequest download routines).
+	errs := make(chan error, 1+concurrentDownloads)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	wg := sync.WaitGroup{}
+
+	// In one routine, get all IDs matching the filter and pass them into ids.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// If limit is too high the request can time out, so limit each call to a maximum value of 100.
+		var limit int
+		if concurrentDownloads > 100 {
+			limit = 100
+		} else {
+			limit = concurrentDownloads
+		}
+		if err := getMatchingBinaryIDs(ctx, c.dataClient, filter, ids, limit); err != nil {
+			errs <- err
+			cancel()
+		}
+	}()
+
+	// In parallel, read from ids and download the binary for each id in batches of defaultConcurrentDownloads.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var nextID string
+		var done bool
+		numFilesDownloaded := &atomic.Int32{}
+		downloadWG := sync.WaitGroup{}
+		for {
+			for i := 0; i < concurrentDownloads; i++ {
+				if err := ctx.Err(); err != nil {
+					errs <- err
+					cancel()
+					done = true
+					break
+				}
+
+				nextID = <-ids
+
+				// If nextID is zero value, the channel has been closed and there are no more IDs to be read.
+				if nextID == "" {
+					done = true
+					break
+				}
+
+				downloadWG.Add(1)
+				go func(id string) {
+					defer downloadWG.Done()
+					err := downloadBinary(ctx, c.dataClient, dst, id)
+					if err != nil {
+						errs <- err
+						cancel()
+						done = true
+					}
+					numFilesDownloaded.Add(1)
+				}(nextID)
+			}
+			downloadWG.Wait()
+			if numFilesDownloaded.Load()%logEveryN == 0 {
+				fmt.Fprintf(c.c.App.Writer, "downloaded %d files\n", numFilesDownloaded.Load())
+			}
+			if done {
 				break
 			}
 		}
-		if err != nil {
-			return errors.Wrapf(err, "received error from server")
+		if numFilesDownloaded.Load()%logEveryN != 0 {
+			fmt.Fprintf(c.c.App.Writer, "downloaded %d files to %s\n", numFilesDownloaded.Load(), dst)
```

Comment:
nit: I find it helpful as a user to have (downloaded  # files / total files) as a indication of what's done and what's still left to go